### PR TITLE
Perl_sv_setsv_flags - IV/NV & cold code optimisation

### PR DESCRIPTION
**Pull request by richardleach**
`Perl_sv_setsv_flags` is one of the hottest functions in the interpreter,
at least when looking at the coverage from running the test harness.

This PR basically does two things:
* The "fast number" code early in the function now handles one SV being
  an IV and the other a NV, if both are "bodyless" types. This has the
  upshot that an IV will only be upgraded to an NV, not a PVNV, to store
  an NV value. An NV will be downgraded to an IV, not upgraded to a PVIV,
  to store an IV value. This saves having to allocate/free actual bodies.
  
  Having done this, subsequent code paths later in the function are
  rendered completely unreachable and have been excised.
  
  These changes should be transparent to most Perl users, only anyone who
  is actually looking at SV types - presumably in test code - might notice.
  
* The croak-ing code within extremely cold fail-safe code paths has been
  moved out into a helper function. (I'd have put this in cold.c if we
  had one.) This new function need not be optimised.
  
  This change should be entirely transparent.
---------------------------------------------------------------------------------
* This set of changes requires a perldelta entry, and it is included.

### CODE REVIEW

**Comment by richardleach** [comment]
Note:I tried to create a fast path for PV->PV assignments, the most popular path after the fast NULL/IV/NV code. It was easy enough to make modifications to reduce the number of instructions and branches, but the number of actual cycles reported by _perf_ either stayed identical or increased!  

This seemed to be due to much greater front-end and back-end stalls. That could be down to coincidental unfavourable alignment, so might be something to revisit,

Here's a section of the gcov output produced by running `make test` on a patched gcov build , with comments for context.

```
264440315: 4304:    case SVt_PV: /* Hit 67% of the time when running Perl's test suite */
264440315: 4305:        if (LIKELY(dtype == SVt_PV)) { ~70% of the time in this _case_
186543132: 4306:          fast_pv:
245826680: 4307:            sflags = SvFLAGS(ssv);
245826680: 4308:            if ( (sflags & (SVf_ROK|SVp_POK)) == SVp_POK) /*taken >90% of the time >
245201806: 4309:                goto pv_pok;
 77897183: 4310:        } else if (dtype < SVt_PV) { ~22% of the time in this _case_
 59283548: 4311:            sv_upgrade(dsv, SVt_PV);
 59283548: 4312:            goto fast_pv;
        -: 4313:        }
 19238509: 4314:        break; 7% of the time in this _case_
```

File: sv.c
Comment (tonycoz) [inline sv.c]: This was already added to `proto.h` when you added it to `embed.fnc`.

File: sv.c
Comment (tonycoz) [inline sv.c]: This and the other calls below should be sv_setsv_flags_croak() not S_sv_setsv_flags_croak() for threaded builds to build.

Comment (richardleach) [reply sv.c]: Thanks, tidying up now.

File: sv.c
Comment (tonycoz) [inline sv.c:17721]: I'm not so sure about the function name, it sounds like a version of sv_setsv_flags() that croaks in some circumstances rather than a function sv_setsv_flags() calls when it needs to croak.

maybe croak_for_sv_setsv() (or with _flags) but better ideas welcome.

Comment (richardleach) [reply sv.c:17721]: Yes, good point. Will sleep on it.

**Comment by tonycoz** [comment]
> Note:I tried to create a fast path for PV->PV assignments, the most popular path after the fast NULL/IV/NV code. It was easy enough to make modifications to reduce the number of instructions and branches, but the number of actual cycles reported by perf either stayed identical or increased!

With regards to the performance questions, I believe Intel vTune is intended for answering these types of question, and unlike years ago is part of the free oneAPI toolkit.  I haven't had a need for it myself (yet) but it might worth trying here if you have the time/inclination.

File: sv.c
Comment (bulk88) [inline sv.c:17732]: Switch as many Perl_croak() to Perl_croak_nocontext() as possible if its in the same C fnc. This is panic code and almost unreachable. Drops all the copy/mov/push cpu instructions passing my_perl.

File: sv.c
Comment (bulk88) [inline sv.c]: factor out reading global PL_op and very likely fn call inside OP_DESC to a root scope instead 2 copies in 2 branches

File: sv.c
Comment (bulk88) [inline sv.c:4337]: I'm not reading the original full .c so maybe Im making mistakes. SVt_IV is always bodyless,   How is this statement even reached with both bodyless NV and area body big NV builds? ```sv_upgrade(dsv, SVt_IV);``` only makes sense for cpan code for public API reasons,

1st problem why is this a fn call to sv_upgrade() instead of bodyless tricks from NULL to IV?

why does this conditional even exist? wasn't it caught above?

File: sv.c
Comment (bulk88) [inline sv.c:4280]: You are baking in the integers behind the SvTYPEs. Needs a static assert somewhere to protect your bitwise math  tricks/optimizations against the SvTYPE flags changing.  There are no guarantees SVIV stays as 1 and SVNV as 2, for eternity. I wrote a branch some years ago that removed the SVT_NULL flag, and SVT_IV is type 0, and but it wasn't accepted, or too many APIText.xs and B.xs  .t'es to fix (100s, extreme limit 1K test fails). The test failures were all str eq'es against deparse or devel::peak debugging console output. No SEGVs.

Seems pointless to deliver fresh empty SV heads, with a NULL for sv_any/sv_body * instead of SVT_IV, only for core and cpan XS/C code to maybe do SV_UPGRADE() or sv_setiv() later. The "-" op to upgrade SVT_NULL to SVT_IV only 1 extra CPU OP op. 2 ops on ARM/RISC.

Again, there are no guarantees SVIV stays as 1 and SVNV as 2, for eternity. I suspect in 2-5 years a tiny NV and 128b NV type might appear 1 NV type. Not a 100% QUADMATH perl. But 2 NV types. Non-apple ARM and especially Apple ARM, are very aggressive on "fork-lift"ing their ARM instruction set every 9 months or 18/24 months. "ABI" barely exist in ARM/smartphone world.  Even for ARM server/cloud/desktop type machines with GNOME Linux, CPU/Hardware vendor-lock in is the market strategy.

A brand new BSD license (secretly commercial) JSON parser written for exactly THAT 1 CPU model number is needed to avoid a 75% speed drop on that ARM vs a GNU/FSF JSON parser on x64, same mhz/same MIPS/same superficial FooMark2000 scores. And the GNU/GPL old JSON library crawls on the brand new ARM. So always think ahead, your upstream legacy libraries can suddenly drop dead performance wise or go QEMU slow in ARM world, and another "best practice" FOSS library becomes popular. Official Zlib vs CPU specific forks is notorious problem.

A Perl API future outlook hazard. SvTYPE is currently UNSIGNED. Over time, any code base in C or psuedo-JIT C, UNSIGNED fields become signed, because "sign test flag" and "zero test flag" come free, and automatic, with every ```<< >> | & ~ + -```. See V8's 31 bit SMI integers rational.

In a Perl specific API, nobody core dev right now can tell, if a patch comes in the future, justifying "negative" type code/product codes/feature flags, or a ```x < 0 ||  x >0```  where 0 in the middle or lower 1/3 or upper 1/3 of the all "product codes". Because it was determined that there is a very crucial, very high frequency, loose behavior classification test, where writing a U8 const operand like 4 or 7 or 11, needs to be turned into a 0 test or sign bit used to avoid encoding U8 constants in machine code over and over, and product code 11 out of 28, needs to be made 0, and  product code <11 become negative U8s, to utilize the sign flag that comes free with every CPU op.

File: sv.c
Comment (bulk88) [reply sv.c:17721]: Agree with TC, keep the identifiers as Perl_croak_something() and croak_something(), so in alpha-sort in GUI in a "report tool" or IDE, and a grep,  all the  perl fatal error support fncs are sitting are quickly visible to a core dev.

File: sv.c
Comment (Leont) [reply sv.c:17732]: Why would we want that? We don't do that with any other the many other PerlAPI calls either.

File: sv.c
Comment (richardleach) [reply sv.c:4337]: Yeah, this existing code is a bit curiously written, presumable for historical reasons. (Maybe it predates bodyless IVs? And RVs no longer having their own type?)

Prior to this PR, regardless of NV size, any NULL will be upgraded to an IV via bodyless tricks, and this will happen in the special "bodyless" casing at the top of the function.

With bodyless NV:
* Prior to this PR, the special "bodyless" casing didn't handle a source IV and a destination NV (and vice versa), so this code was reachable.
* With this PR, this code is no longer reachable - and such builds will fail the `#if` test.

With arena body NVs, this statement is still reachable, because the special bodyless casing will only handle NULL/IVs in such builds. However, the code could be more clearly rewritten now as:
 
```
#if NVSIZE > IVSIZE
        if (dtype == SVt_NV)
            sv_upgrade(dsv, SVt_PVIV);
```
And we only get here if `SvROK(ssv)`. I'm unclear on whether a NV is meant to be capable of storing a RV or not. If not (which seems likely), I don't know if the correct thing to do is excise this code or swap the `sv_upgrade` for a `croak`. It's currently still here out of caution for this one potential use-case, basically.

File: sv.c
Comment (richardleach) [reply sv.c:4280]: The top of the function has us covered on the current SvTYPE integers:
```
    STATIC_ASSERT_STMT(SVt_NULL == 0);
    STATIC_ASSERT_STMT(SVt_IV   == 1);
    STATIC_ASSERT_STMT(SVt_NV   == 2);
```

Your point about delivering bodyless `SVt_IV` as a minimum, rather than `SVt_NULL` is an interesting one. I'd have to go spelunking to know if any code is predicated on SVs being of the `SVt_NULL` type tho.

At the current pace of core development, waiting to think about the other points only if and when they arrive seems reasonable.

File: sv.c
Comment (bulk88) [reply sv.c:17732]: Removing the my_perl arg on threaded perl from each croak() call site, drops 4-7 bytes of machine code, probable 8 or 12 bytes on ARM64. do it on 20 lines/call points. Now 100 bytes is gone from libperl, through a regexp substitute and very little thinking/time required by the author for a cheap fast (dev human time) little perf boost.

Assert branches/mem corruption panics printing branches are never perf critical, and almost unreachable for production perl code. And "perl code" calling PP ```die()``` does pass my_perl ptr through the ```Perl_pp_die()->Perl_die_sv()->Perl_croak_sv()->Perl_die_unwind()->longjmp()``` call chain, so we don't strip the my_perl ptr through PP die, Ive check in the past for bad Perl_get_contexts() there, PP_die() has no low level Perl specific perf issues IMO.  But for individual C API core die()/panics() format string, they are pretty heavy in Perl C. Every C level argument minimized to the the format string is better.

Excellent debugging info has a place, but perl doesnt use int error codes or error objects passed internally for its panic/semi asset messages. Most C/C++ code doesn't keep the ASCII text string support/debug code anywhere near average runtime code. The end user must explicitly pass that int error code or error object to a "error exception decoder" framework in another cold .o file in that app if they want an ASCII string description, 

Perl core traditionally keeps its fatal "text error formatter" code, right inside a C for loop, and perl often drops 4-8 pointers worth of "push &some_const_string_U32" format strings in machine code, and that overhead builds up if 2 64 bytes/8 ptr cache lines are 80% cold format C string args/getters funcs/misc, and 20% production statements on the edges of that 128 byte unit. Threaded perl is another layer of runtime overhead, but if the Perl C API, is designed a little bit smarter (PerlEnv vtable API needs a diet). Threaded perl/IMPLICT_CONTEXT/my_perl pointer builds, are just the "this" pointer in C++ machine code wise anywhere else. I've never seen anywhere in soft dev community complaints where someone wants to "optimize away" the ```this``` var in C/C++.

My point is CPU has to skip over format string machine code, and those branches never execute unless the process is about to SEGV, and nobody trap/resume execution after a ```bizzare copy of``` error. Aslong as ```Perl_pp_die()``` PP support written in C code, passes around my_perl, nobody can harm that their PP perl benchmark that measures units of "1000 exceptions per second" or "1000 die();s per second" was made worse. Perl_croak_nocontext() was originally a back compat fix for var args/CPreProc doesnt do var args, but IMO Perl_croak_nocontext() is little known tool that should be used more for reducing the "perl is slow" debate.

The user at runtime can spend an extra 2-4 microseconds and go back into the OS's TLS API and get back the my_perl ptr (Perl_get_context()) to finishing printing to stderr/expanding format strings, and the process is about to exit(1) or SEGV in a couple milliseconds anyways. 4 microseconds or 10 us of CPU or wall time can be spent, on a proc about to exit, in exchange for a permanent runtime benefit in primary execution flow/hot or warm code,  to all other global users.

File: sv.c
Comment (bulk88) [reply sv.c:4337]: I dont remember from 5 years ago enough, but I think P5P opinion is to never implement SV body downgrades.  SVRV is long gone as an independent type code and its just a SVIV with a flag. Perhaps 5.6/5.8 era perl bless related struct fields got moved from SVRV out wrapper to SVMG real obj, and SVRV lost its purpose. Someone could argue Perl's ```SvTYPE()``` is a solid U8 of metadata, and needs to represent "defined contents" of the SV more than ```SvTYPE()```'s current malloc book keeping purpose.  Perl SV head U32 or U24 bitfield, encode so much metadata, and the behaviors modeled are not` ``adjacent``` (fuzzy word) in the bitfield,  so U32->U8 split bitwise tricks are hard. So P5 core dev cant do ``` for(i=6;i;i--) {sv_flags = sv_flags >>1; if(){}}``` type things. other than IPNR/IPNRp-Ok() in byte 2 which is linear.

A certain perl person, not me, not naming them, did argue with me that P5P keeps their metadata too many pointers/layers deep. And totally wasted the "gift" of the SV head's 4th slot. This person wants to see var len malloced C structs at the start of HvARRAY and AvARRAY and PV buffer/VLAs, and not the current and day 1 perl 5 setup of array ptrs being pure array ptrs.

PV HEK and UTF8PV sneak in metadata into the "array", but these 3 APIs keep it on the end at random offsets. its not a struct.

He also proposed deleting field 4 in the SV head, and ```->sv_any``` becomes malloced, and toss the legacy P5 arena API except for SV heads.

I wouldnt vote yes if I had to make an official vote with his design proposals above, but this person is correct IMO that the HV * layout and HE*  has way too many ptr derefs. Ive seen at YAPCs, commercial grade code, that does PP Arrays with PP const subs labels for array indexes. Its not a blog post semi-joke anymore.

Blessed arrays with integer labeled methods, yes there is commercial Perl code with that design (C++ objects in PP) running, but its a minority, profilers would show I/O, SQL DBs, OSes, and ping time matter more.

LVALUE tie/LVALUE magic vs SV .h macros vs perlguts POD is most difficult part of the SV scalar half of the type system to learn or debug. LVALUE set magic firing on lvalue assignment from an XSUB. Plus LVALUE type is rarely seen I think, most apps and most scalar SVs on planet earth only peak at SVt_PVMG, and SVt_PVLV is IV NV PV buffered string type, but it sits out of place in the type system.

SVGV is below SVPVLV. So if there will be bugs related to P5's exactly architecture PP types in implemented in C types, they will happen in SVPVLV type, CPAN, p5p, or private XS code. Useless SvUPGRADE() calls to type PVLV is thought experiment and how many asserts, SEGVs, and real bugs come up.

File: sv.c
Comment (bulk88) [reply sv.c:4280]: I did a ``` #define``` of SVt_NULL to SVt_IV to get core and average river CPAN XS to compile. There is window for too smart CPAN XS code to fail on such a build. SvUPGRADE doesn't change type numbers anymore for no-SVt_NULL perl. And ALOT of SvUPGRADE() calls optimize away by the CC, but too fancy XS code will spin loop hang or SEGV when ```if(type > type)``` code was written, and suddenly only ``if(type `>= type)``` will pass the test on no-SVt_NULL. My main goal in 2016 for no-SVt_NULL perl was SVIV int math speed rational and int math PP perl benchmarks. And SvUPGRADE on type 0 (new SVT_IV) was guaranteed removed by CPreProc in core and CPAN in "1 shot", or bomb blast.  But breakage, umm, early days of ithreads bad? (nuclear fallout) or removing "." from @INC bad? IDK.

Its possible for no-SVt_NULL perl to exist, it completed a make test without a SEGV (but had maybe 1000 test fails). Which means it could be a stable perl one day with enough depreciation to CPAN river or natural abandonment of CPAN XS modules. Fixing the core .t'es was too big of a mountain for me weeks/months wise with zero gain for LOC involved.

Step debugging and refactor specific "pp_foo()" or specific C funcs I catch at runtime personally, felt like better use of my time vs the global nuclear bomb blast of the CC eliminating all SVUPGRADE to SVIV fnc calls and possible major CPAN breakage and outrage, and the work put into patch would get it reverted quick after being declared stable and shipping initially. risk reward.

File: sv.c
Comment (richardleach) [reply sv.c:17721]: Ok, I went with `S_croak_sv_setsv_flags`. It seemed a reasonable fit with the existing `croak_*` names.

File: sv.c
Comment (richardleach) [reply sv.c]: Done & pushed.

**Comment by tonycoz** [comment]
One other thing here: this is apparently a performance optimization, the results of benchmarks would be useful, and the conditions they were measured under.

**Comment by richardleach** [comment]
Thanks for the reminder, @tonycoz. I measured a few different scenarios using `perf stat` against blead and with this PR applied ("patched"). Both builds had `ivsize=8` and `nvsize=8`, so both IV and NV types used the "bodyless" mechanisms. There was minor noise between runs, as usual, so multiple runs were done to account for this.

No measurable change was expected or observed on:
* `my $x = 1; for (1..100_000_000){ $x = 1; }`     (**IV into IV assignment**)
* `my $x = 1.1; for (1..100_000_000){ $x = 1.1; }` (**NV into NV assignment**)

There seemed to be a tiny improvement to the following, but nothing significant:
* `my $x = ""; for (1..100_000_000){ $x = "Just another Perl Hacker" }` (**PV into PV**)

(The fast-NULL/IV/NV block at the top of the function, plus the PV->PV path, account
for ~70% of calls to Perl_sv_setsv_flags, according to gcov when running the test harness.)

Noticeable changes were expected and observed when blead upgraded to a type
with an allocated body, but the patched version did not.

* `my $x = 1.1; for (1..100_000_000){ $x = 1.1; $x = 1;}`  (**alternative IV/NV assignments**)

blead:
```
          2,897.24 msec task-clock                       #    0.993 CPUs utilized          
                 4      context-switches                 #    1.381 /sec                   
                 0      cpu-migrations                   #    0.000 /sec                   
               189      page-faults                      #   65.234 /sec                   
    12,670,977,088      cycles                           #    4.373 GHz                    
        43,543,251      stalled-cycles-frontend          #    0.34% frontend cycles idle   
           910,784      stalled-cycles-backend           #    0.01% backend cycles idle    
    42,807,329,051      instructions                     #    3.38  insn per cycle         
                                                  #    0.00  stalled cycles per insn
     9,001,709,241      branches                         #    3.107 G/sec        
```          
                 
patched:
```
          2,575.52 msec task-clock                       #    0.988 CPUs utilized          
                 5      context-switches                 #    1.941 /sec                   
                 0      cpu-migrations                   #    0.000 /sec                   
               186      page-faults                      #   72.218 /sec                   
    11,289,082,929      cycles                           #    4.383 GHz                    
        17,195,387      stalled-cycles-frontend          #    0.15% frontend cycles idle   
         1,761,258      stalled-cycles-backend           #    0.02% backend cycles idle    
    41,206,945,181      instructions                     #    3.65  insn per cycle         
                                                  #    0.00  stalled cycles per insn
     8,501,624,007      branches                         #    3.301 G/sec
```

* `for (1..100_000_000) { my @nums = (1); $nums[0] = 1.1 }` (**setting an IV, assigning an NV, freeing the SV**)

blead:

```
          7,901.93 msec task-clock                       #    0.998 CPUs utilized          
                 5      context-switches                 #    0.633 /sec                   
                 0      cpu-migrations                   #    0.000 /sec                   
               189      page-faults                      #   23.918 /sec                   
    36,490,011,259      cycles                           #    4.618 GHz                    
       200,534,959      stalled-cycles-frontend          #    0.55% frontend cycles idle   
         2,269,549      stalled-cycles-backend           #    0.01% backend cycles idle    
   109,313,808,352      instructions                     #    3.00  insn per cycle         
                                                  #    0.00  stalled cycles per insn
    21,903,161,401      branches                         #    2.772 G/sec 
```
    
patched:
```
          6,445.95 msec task-clock                       #    0.993 CPUs utilized          
                43      context-switches                 #    6.671 /sec                   
                 0      cpu-migrations                   #    0.000 /sec                   
               187      page-faults                      #   29.010 /sec                   
    28,437,333,228      cycles                           #    4.412 GHz                    
       282,558,412      stalled-cycles-frontend          #    0.99% frontend cycles idle   
           100,532      stalled-cycles-backend           #    0.00% backend cycles idle    
    94,612,222,136      instructions                     #    3.33  insn per cycle         
                                                  #    0.00  stalled cycles per insn
    18,802,825,054      branches                         #    2.917 G/sec
```

* `my @nums = (1) x 100_000_000; for (1..100_000_000) { $nums[$_] = 1.1 }` (**NVs into a large array of IVs**)

blead:
```
         12,682.24 msec task-clock                       #    0.900 CPUs utilized          
            19,272      context-switches                 #    1.520 K/sec                  
                 0      cpu-migrations                   #    0.000 /sec                   
         1,674,061      page-faults                      #  132.000 K/sec                  
    47,687,795,920      cycles                           #    3.760 GHz                    
       356,547,911      stalled-cycles-frontend          #    0.75% frontend cycles idle   
     1,659,808,928      stalled-cycles-backend           #    3.48% backend cycles idle    
    90,426,683,028      instructions                     #    1.90  insn per cycle         
                                                  #    0.02  stalled cycles per insn
    19,795,775,732      branches                         #    1.561 G/sec
```
patched:
```
          5,468.95 msec task-clock                       #    0.995 CPUs utilized          
                67      context-switches                 #   12.251 /sec                   
                 0      cpu-migrations                   #    0.000 /sec                   
           595,020      page-faults                      #  108.800 K/sec                  
    24,397,472,622      cycles                           #    4.461 GHz                    
        73,737,773      stalled-cycles-frontend          #    0.30% frontend cycles idle   
     1,388,156,279      stalled-cycles-backend           #    5.69% backend cycles idle    
    67,612,961,402      instructions                     #    2.77  insn per cycle         
                                                  #    0.02  stalled cycles per insn
    14,661,104,730      branches                         #    2.681 G/sec

```
`/usr/bin/time -v` was used to measure the Maximum resident set size (kbytes) for the above large array allocation runs: 

blead: 7738900 kbytes
patched: 4871548 kbytes

**Comment by richardleach** [comment]
> > Note:I tried to create a fast path for PV->PV assignments, the most popular path after the fast NULL/IV/NV code. It was easy enough to make modifications to reduce the number of instructions and branches, but the number of actual cycles reported by perf either stayed identical or increased!
> 
> With regards to the performance questions, I believe Intel vTune is intended for answering these types of question, and unlike years ago is part of the free oneAPI toolkit. I haven't had a need for it myself (yet) but it might worth trying here if you have the time/inclination.

I'll add doing this to my list for 2025. :)

File: sv.c
Comment (richardleach) [reply sv.c:17732]: > Switch as many Perl_croak() to Perl_croak_nocontext() as possible

I think that if we want to do this, it should be a separate PR.

File: sv.c
Comment (richardleach) [reply sv.c:4337]: I too am interested in the potential for alternatives to the current type bodies (or even doing "fat heads" for e.g. AV and HV"), but not as part of this PR.

File: sv.c
Comment (richardleach) [reply sv.c:4280]: Also a separate PR if we want to do this.

**Comment by richardleach** [comment]
Rebased. Please let me know if there are any further changes needed before it is good to merge.

**Review by tonycoz — APPROVED** [review]

**Pull request merged**
